### PR TITLE
1029 create custom handling for hi goodtimes jobs

### LINKED ISSUE #1029: Create custom handling for hi goodtimes jobs
### Algorithm Description:

Hi Goodtimes needs several repointings worth of data to run. This ticket will implement custom handling of Hi Goodtimes.

### PR DESCRIPTION
**Pull request by tmplummer**
# Change Summary

## Overview
                                                                                                                                                                                                                                                                
  This PR implements custom handling for Hi Goodtimes jobs, which require L1B DE data from multiple repoints (N=7 total, including the target repoint).

## Key Changes

  1. Batch Starter (batch_starter.py)

  - Added special handling when a Hi L1B DE file triggers Hi Goodtimes jobs
  - When triggered, expands to multiple target repoints in range [T-N+1, T+N-1] where T is the trigger repoint and N=7
  - Each target repoint gets its own job submission with normal dependency checking

  2. Dependency Module (dependency.py)

  New Configuration:
  - HI_GOODTIMES_NUM_NEAREST_REPOINTS = 7 - Total repoints needed for Goodtimes processing

  New Functions:
  - get_hi_goodtimes_target_repoints() - Returns target repoints in range [T-N+1, T+N-1]
  - _extend_hi_goodtimes_l1b_de_dependencies() - Extends L1B DE dependencies to include N nearest repoints
  - get_n_nearest_files_by_repoint() - Generic function to find N nearest files by repoint number
  - get_n_nearest_files_by_date() - Generic function to find N nearest files by date
  - _get_available_repoints() - Query helper for available repoints in database
  - _get_inprogress_repoints() - Query helper for repoints with INPROGRESS jobs
  - _get_inprogress_dates() - Query helper for dates with INPROGRESS jobs
  - _check_pointing_exists() - Check if a pointing exists in the pointing table

  Modified Functions:
  - get_files() - Now accepts repoint as either int or list[int] to query multiple repoints
  - get_upstream_dependency_inputs() - Updated to accept list of repoints
  - get_jobs() - Calls _extend_hi_goodtimes_l1b_de_dependencies() for Hi Goodtimes jobs

  Hi Goodtimes Job Logic:
  1. When processing a Goodtimes job for target repoint T, find N-1 nearest L1B DE repoints
  2. Skip job submission if any of the N nearest repoints have INPROGRESS L1B DE jobs
  3. If fewer than N/2 future repoints exist, verify pointing T+N-1 exists before proceeding
  4. Extend the L1B DE dependency to include files from all N repoints

  Test Coverage

  - Unit tests for all new helper functions (_get_inprogress_repoints, _get_inprogress_dates, _check_pointing_exists, get_hi_goodtimes_target_repoints)
  - Unit tests for get_n_nearest_files_by_repoint and get_n_nearest_files_by_date
  - Integration tests for Hi Goodtimes multi-repoint handling
  - Tests for INPROGRESS job skipping behavior
  - Tests for edge cases (only past repoints, only future repoints)

Closes: #1029

### CODE REVIEW

File: tests/lambda_endpoints/test_batch_starter.py [hunk 1]
```diff
@@ -2292,3 +2292,125 @@ def test_determine_job_version_spacecraft(session):
...
     assert version == "v003"
+
+
+@patch(
+    "sds_data_manager.lambda_code.SDSCode.pipeline_lambdas.batch_starter.dependency.get_dependencies"
+)
+@patch(
+    "sds_data_manager.lambda_code.SDSCode.pipeline_lambdas.batch_starter.submit_all_jobs"
+)
+def test_hi_goodtimes_multi_repoint_trigger(
```
Comment (lacoak21) [inline tests/lambda_endpoints/test_batch_starter.py:2339]: Might be worth it to add a test where the job is not triggered due to one of the l1b de repoints missing.

File: tests/lambda_endpoints/test_dependency_api.py [hunk 2]
```diff
@@ -780,6 +875,121 @@ def test_get_files_max_version(session):
...
+    return_value=[
+        {
+            "data_source": "hi",
+            "data_type": "l1b",
+            "descriptor": "45sensor-de",
+            "relationship": "HARD",
+        }
+    ],
+)
+def test_get_jobs_hi_goodtimes_multi_repoint(
```
Comment (lacoak21) [inline tests/lambda_endpoints/test_dependency_api.py:1012]: What would happen if you called it like 

`    science_files = get_files(
        session,
        dependency=dep,
        start_date=datetime(2024, 1, 1),
        end_date=datetime(2024, 1, 2),
        repoint=[1, 2],
    )`

And there was only data for repoint 1?

**Review by lacoak21 — APPROVED** [review]
Looks good! Just a couple of questions related to missing repointing data.

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py [hunk 3]
```diff
@@ -1233,11 +1243,30 @@ def get_jobs(
     start_date = datetime.strptime(start_date, "%Y%m%d")
     end_date = datetime.strptime(end_date, "%Y%m%d")
 
+    # Special handling for Hi Goodtimes - needs L1B DE from multiple repoints
+    # Pass a list of repoints instead of a single repoint
+    repoint_param = repoint
+    if (
```
Comment (maxinelasp) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py]: This is a duplicate if-statement from batch starter, is it possible to combine the checks into one place?

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py [hunk 4]
```diff
@@ -807,15 +808,51 @@ def s3_processing_event(session, events):
...
+                    repoint + dependency.HI_GOODTIMES_NUM_PAST_REPOINTS + 1,
+                ):
+                    logger.info(
+                        f"Submitting Hi Goodtimes job for repoint {target_repoint} "
+                        f"(triggered by repoint {repoint} file)"
+                    )
+                    submit_all_jobs(
+                        session,
+                        job,
+                        trigger_start_time,
```
Comment (maxinelasp) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py:869]: Does the start and end time need to be changed to match repoint?

**Review by maxinelasp — APPROVED** [review]

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py (hunk 4 above)
Comment (tmplummer) [reply sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py:869]: Not here. This is just submitting a potential job for each of the repoints that the trigger file would be used in. It is in the dependency code where the start/end time need to get updated when gathering upstream dependencies for a job.

File: tests/lambda_endpoints/test_batch_starter.py (hunk 1 above)
Comment (tmplummer) [reply tests/lambda_endpoints/test_batch_starter.py:2339]: Good call. I started working on this and quickly realized that there is no checking that the coverage by the retrieved dependencies is correct. I opened another PR to implement that checking. 
https://github.com/IMAP-Science-Operations-Center/sds-data-manager/pull/1041

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py [hunk 5]
```diff
@@ -1312,12 +1608,15 @@ def get_files(
...
+        # If repoint is provided, filter by repointing number(s)
         if (
             repoint is not None
             and dependency["data_source"] in REPOINT_DEPENDENT_INSTRUMENTS
         ):
-            type_specific_conditions.append(table.repointing == repoint)
+            if isinstance(repoint, list):
+                type_specific_conditions.append(table.repointing.in_(repoint))
+            else:
+                type_specific_conditions.append(table.repointing == repoint)
```
[the remark this answers is not in the document]
Comment (tmplummer) [reply sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py]: If each repoint corresponds to a distinct time period, they would naturally have different start dates. So the current get_files logic of grouping by start_date to find max version is correct for our use case.

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py [hunk 6]
```diff
@@ -819,16 +820,55 @@ def s3_processing_event(session, events):
...
+                and file_obj.descriptor.endswith("-de")
             )
 
+            # Special handling: When Hi L1B DE triggers Hi Goodtimes,
+            # expand to multiple target repoints
+            if (
+                trigger_is_hi_l1b_de
+                and repoint is not None
+                and job["data_source"] == "hi"
+                and job["data_type"] == "l1c"
```
Comment (tmplummer) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py]: I think we decided that the ENA goodtimes would be l1b products.
```suggestion
                and job["data_type"] == "l1b"
```

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py [hunk 7]
```diff
@@ -1463,5 +1951,26 @@ def get_jobs(
         )
         return None
 
+    # Special handling for Hi Goodtimes - extend L1B DE to N nearest repoints
+    if (
+        data_type == "l1c"
```
Comment (tmplummer) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py]: We decided to make Goodtimes L1B
```suggestion
        data_type == "l1b"
```

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py [hunk 8]
```diff
@@ -1076,14 +1096,20 @@ def get_upstream_dependency_inputs(
...
 
     Returns
     -------
     ProcessingInputCollection
         Dependency files that can include Ancillary, SPICE, or Science inputs.
     """
     dependency_inputs = processing_input.ProcessingInputCollection()
-    with db.Session() as session:
+    # Use provided session or create a new one
+    session_context = (
```
Comment (tech3371) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py]: creating new db session if not passed in here makes sense.

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py [hunk 9]
```diff
@@ -819,16 +820,55 @@ def s3_processing_event(session, events):
...
+                target_repoints = dependency.get_hi_goodtimes_target_repoints(
+                    trigger_repoint=repoint,
+                )
+
+                for target_repoint in target_repoints:
+                    logger.info(
+                        f"Submitting Hi Goodtimes job for repoint {target_repoint} "
+                        f"(triggered by repoint {repoint} file)"
+                    )
+                    submit_all_jobs(
```
Comment (laspsandoval) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py:866]: I think this answers my question from yesterday. 

  - Reprocess L1B DE specifically -> s3_processing_event fires and fan-out happens normally.                                                                                   
  - Reprocess Goodtimes directly goes straight to submit_all_jobs, bypasses s3_processing_event entirely -> no fan-out.

**Review by laspsandoval — APPROVED** [review]

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py [hunk 10]
```diff
@@ -819,16 +820,55 @@ def s3_processing_event(session, events):
...
-                session,
-                job,
-                trigger_start_time,
-                trigger_end_time,
-                repoint,
-                calculate_crids,
-                filter_dependencies,
+
+            # Check if trigger file is Hi L1B DE
+            trigger_is_hi_l1b_de = (
```
Comment (lacoak21) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py:839]: Is l1b de the only dependency? Im just wondering whether you needed to handle a case where you had to reprocess goodtimes because a SPICE kernel got updated.

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py [hunk 11]
```diff
@@ -1360,6 +1700,176 @@ def get_files(
     return records
 
 
+def get_n_nearest_files_by_repoint(
```
Comment (lacoak21) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py:1703]: @subagonsouth I reviewed part of this but had to stop here. I will resume in the morning.

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py [hunk 12]
```diff
@@ -1255,12 +1279,320 @@ def yyyymmdd_to_seconds_since_j2000(
...
+    l1b_de_records = get_n_nearest_files_by_repoint(
+        session,
+        l1b_de_dep,
+        start_date,
+        end_date,
+        num_nearest,
+        repoint,
+        skip_if_inprogress=True,
+    )
+    if l1b_de_records is None:
```
Comment (lacoak21) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py:1513]: Uhoh I think you are right that other jobs dont check for upstream jobs in progress. The CRID check only handles it when an upstream , upstream job is still processing. Wow how could I have missed this :(. Ill make sure to keep this in mind when reviewing Tenzin's refactor.

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py [hunk 13]
```diff
@@ -1360,6 +1700,176 @@ def get_files(
...
+        logger.info(f"Target repoint {repoint} not found for {dependency}")
+        return []
+
+    # Remove target, sort by distance then repoint, take N nearest
+    other_repoints = all_repoints[all_repoints != repoint]
+    if len(other_repoints) == 0:
+        return []
+
+    distances = np.abs(other_repoints - repoint)
+    sort_indices = np.lexsort((other_repoints, distances))
```
Comment (lacoak21) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py:1765]: Whoa np.lexsort is so useful.

File: sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py [hunk 14]
```diff
@@ -1448,20 +1958,49 @@ def get_jobs(
...
-        end_date=end_date,
-        repoint=repoint,
-        calculate_crids=calculate_crids,
-        get_spice=get_spice,
-        require_coverage=require_coverage,
-    )
-    if upstream_dependencies_output is None:
-        logger.info(
-            f"No dependencies found for {start_date=} - {end_date=}: {dependencies}"
+    # Use a single session for all database operations
```
Comment (lacoak21) [inline sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/dependency.py:1961]: Good call.

**Review by lacoak21 — APPROVED** [review]
This looks great Tim. Have you tested in dev out of curiosity?
